feature: add a v2 crd generator based on jackson

shawkins opened the pull request:
## Description
closes: #5948

also removes v1beta1 from the new code

This is not yet ready to merge. For the tests to pass we need to:
- confirm the JsonFormat shape behavior from https://github.com/fabric8io/kubernetes-client/discussions/5942#discussioncomment-9214458 - for now the types were changed to string
- The required list was previously unordered, now it is. @andreaTP are there any objections to making that the new expected behavior?
- The printer column type is taken directly from the property - but in the test it's expecting a string instead of an array type for one of the columns. What is the expectation here - is string required, or was that just a limitation of the previous logic? cc @andreaTP 
- It now better understands the anygetter / setter logic, so more properties are marked as preserve unknown causing another test to fail. Thinking about this more, this one is not clear cut. This is coming by the way of core model objects, which we include additionalProperties on as a way of letting them be foward compatible - it's not as a way to indicate what the schema actually should be. It could be using the custom PreseveUnknown annotation is the way to go here, rather than detecting the situation ourselves. cc @andreaTP 

I'll add a changelog entry as well once we're ready.

cc @matteriben @baloo42

## Type of change
<!---
What types of changes does your code introduce? Put an `x` in all the boxes that apply
-->
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [ ] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/main/CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/main/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

<!--
Integration tests (https://github.com/fabric8io/kubernetes-client/tree/master/kubernetes-itests)
Please check integration tests and provide/improve tests if applicable.

Open your PR in Draft mode and verify all of the applicable Checklist items before marking your pull request as ready for review
-->

shawkins commented:
I went with crdv2 as the package name - but that can be changed of course.

baloo42 commented:
> The required list was previously unordered, now it is.
I think we should handle this as a bug for the v1 generator and also fix it. If not, the new approval tests (#5943) would also fail sometimes. The outcome should be always deterministic.

shawkins commented:
@baloo42 I've updated the test for v2 to keep this pr moving forward.

@matteriben based upon https://github.com/fabric8io/kubernetes-client/discussions/5942#discussioncomment-9217399 it sounds like we're moving forward with leaving the type as string for those shape tests, which puts us in a similar situation as sorting the required list - the v1 implementation should be updated to match.

That just leaves two differences (that we currently know of) to account for.

andreaTP commented:
> The required list was previously unordered, now it is.

This is a feature, great!

> was that just a limitation of the previous logic?

IIRC that's it, in any case printer columns are just for human consumption, we can solve any issue at a later stage without a significant risk of breaking user code badly.

> It could be using the custom PreseveUnknown annotation is the way to go here, rather than detecting the situation ourselves.

Looking at how ppl are using things I would not be surprised if we take the opinionated decision to add `x-kubernetes-preserve-unknown` everywhere by default ...
That said, I would not forcefully try to follow the current detection logic and just be sure we document the behavior appropriately.

Currently on the phone and I'll be out until Mon, but let me know if any additional input from me is needed to progress here and I'll try to answer asap.

shawkins commented:
> IIRC that's it, in any case printer columns are just for human consumption, we can solve any issue at a later stage without a significant risk of breaking user code badly.

Ok, to keep this pr moving forward I'll update the results here to expect the actual type.

> That said, I would not forcefully try to follow the current detection logic and just be sure we document the behavior appropriately.

The logic here is a little broader - so I'm wondering if it needs to be more limited. The concern is that the object model's intent with using additional properties is to increase compatibility - not that the real schema of ObjectMeta etc. actually calls for preserve unknown. I'm thinking that we'll either need to just look for the PreserveUnknownFields annotation or let it be user configurable in some other way to elect whether anyGetter / anySetter means preserve unknown fields.

shawkins commented:
Marking as ready for review as all current tests pass. Let me know what needs addressed before this gets committed. This also wires up parallel execution (should likely now default to true)  #5952 and exposes two new options for the crd generator:
- the ability to supply an ObjectMapper (this should be the same instance that is passed to the KubernetesSerialization. If we want to add more of backdoor for exposing the ObjectMapper from an instance of KubernetesSerialization, then we can pass the KubernetesSerialization instead.
- the ability to control whether we implicitly treat anyGetter / anySetter as preserve unknown fields - we will probably want a better story here eventually.

Recap of some other changes:
- it works at the property level, rather than the field level, so some crds may have fields removed - but those weren't actually being used. This is also the reason test code made fields public or used the lombok Data annotation - Jackson doesn't consider a private field without accessors serializable.
- more support for format and pattern based upon what is already supported by jacksonschema. There is still more that could be done here around their validation extension logic
- RawExtension and HasMetadata interfaces will be mapped to embedded resources
- the logic resolve the jackson schema for nested classes only once, then we'll reuse that schema from the ResolvingContext
- the abstraction to support multiple CRD versions has been left in, but could be removed if we want to revisit it later when needed
- also removed the decorators as we can do the same thing more directly with builders.

@metacosm this should be good now to test out with quarkus

cc @matteriben @baloo42 @manusa @andreaTP

manusa commented on crd-generator/api-v2/src/main/java/io/fabric8/crdv2/generator/AbstractCustomResourceHandler.java:
```diff
@@ -0,0 +1,32 @@
+ * This class encapsulates the common behavior between different CRD generation logic. The
+ * intent is that each CRD spec version is implemented as a sub-class of this one.
+ */
+public abstract class AbstractCustomResourceHandler {
```
Is there any reason why this is a class instead of an interface?

It's my understanding that we can have a  `interface CustomResourceHandler` and then an implementation `V1CustomResourceHandler` or `CustomResourceHandlerV1`, right?

shawkins replied:
> Is there any reason why this is a class instead of an interface?

Not really. It used to hold common state and have a common method for dealing with printer columns. My intent was to add back the method for printer columns, but didn't bother yet.

> It's my understanding that we can have a interface CustomResourceHandler and then an implementation V1CustomResourceHandler or CustomResourceHandlerV1, right?

Correct, it seemed fine to leave in the abstraction layer, used here and for AbstractJsonSchema, but it's not really needed at the moment.

manusa commented on crd-generator/api-v2/pom.xml:
```diff
@@ -0,0 +1,81 @@
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+
+  <artifactId>crd-generator-api-v2</artifactId>
```
Note that this can potentially be implemented using Java 11 source (in case this simplifies implementations)

shawkins replied:
I'm not aware of anything that would change off hand.

manusa commented on crd-generator/api-v2/src/main/java/io/fabric8/crdv2/generator/CRDGenerator.java:
```diff
@@ -0,0 +1,281 @@
+  private ObjectMapper objectMapper;
+  private Map<String, CustomResourceInfo> infos;
+
+  // TODO: why not rely on the standard fabric8 yaml mapping
+  public static final ObjectMapper YAML_MAPPER = JsonMapper.builder(new YAMLFactory()
```
Yes, why not?

shawkins replied:
A comment was lost from @metacosm about how this format seemed different. Haven't looked at it more deeply.

shawkins replied:
The two main changes I see are the standard generator includes the explicit start ---, and it quotes string values. The ordering changes aren't really needed as they should be handled by the construction of the crd object.

I agree that we don't need explicit start, the question is does it matter if there's additional quoting? It shouldn't but we can account for one or both of these differences with an KubernetesSerialization.asYaml method that takes some style options - we are just creating the Dump / DumpSettings on every call anyway.

shawkins replied:
Looking at it more, I don't think we should add yaml options. They really aren't needed for this case either. We could just remove this yaml mapper - the only practical issue is that string comparison of the existing crd output yaml won't work.

shawkins replied:
I can't find any reference going back in the code for that usage of a yaml mapper. The combination of the comment and the start marker are still legal yaml, so I'm not sure what the issue could would have been.

metacosm replied:
Can confirm that the start marker is causing issues, in particular with OLM validation, at least.  I haven't tried applying the CRDs to a cluster, yet, though.

metacosm replied:
Applying the CRD to the cluster seems to work OK, as expected.

shawkins replied:
Thanks for checking @metacosm - the explicit start has been removed

metacosm replied:
Stripping the marker solved the issue as expected. Thank you! 
For reference, here's the successful QOSDK build with the changes from this PR: https://github.com/quarkiverse/quarkus-operator-sdk/actions/runs/8903462629/job/24459014817

shawkins replied:
Great, thank you for testing. The only other thing that could change is supplying the ObjectMapper / KuberenetesSerialization so that we can pick up any changes the user may have made to the mapper configuration

cc @manusa

manusa commented on crd-generator/api-v2/src/main/java/io/fabric8/crdv2/generator/CRDInfo.java:
```diff
@@ -0,0 +1,50 @@
+    return crdName;
+  }
+
+  /**
+   * @deprecated Use {@link #getCrdSpecVersion()} instead
+   * @return the CRD spec version
+   */
+  @Deprecated
+  public String getVersion() {
+    return getCrdSpecVersion();
+  }
```
Should we add a deprecated method to the new API?

metacosm replied:
Probably not but I think it's just moving in stages :)

shawkins replied:
No, I'll remove it.

manusa commented on crd-generator/pom.xml:
```diff
@@ -33,17 +33,7 @@
+    <module>api-v2</module>
+    <module>test</module>
   </modules>
-  <!-- TODO: Remove after a fix for #5877 is implemented -->
-  <profiles>
-    <profile>
-      <id>tests</id>
-      <activation>
-        <jdk>(,17]</jdk>
-      </activation>
-      <modules>
-        <module>test</module>
-      </modules>
-    </profile>
-  </profiles>
```
:+1:

baloo42 reviewed (commented):
Tested it with a few internal projects and looks very good so far.

Beside of the printer columns and the ordering of the required properties list I haven't found any differences to v1.

shawkins commented:
> Tested it with a few internal projects and looks very good so far.
> 
> Beside of the printer columns and the ordering of the required properties list I haven't found any differences to v1.

@baloo42 thank you for the code suggestions and double checking the output.

baloo42 commented:
More general feedback:

#### x-kubernetes-embedded-resource
> RawExtension and HasMetadata interfaces will be mapped to embedded resources

:+1: Nice

#### Printer Columns
> The printer column type is taken directly from the property - but in the test it's expecting a string instead of an array type for one of the columns. What is the expectation here - is string required, or was that just a limitation of the previous logic? 

Kubernetes allows for printer columns only the following types:
boolean,date,integer,number,string

I would suggest to silently filter printer colums on other types. At the moment it produces an invalid CRD.

"Arrays" are only supported if we generate something like the following, which would be an own story:

```
JSONPath: ".spec.hardware.devices.dynamicDirectPathIODevices[*].deviceID"
type: string
```
(the target property must by one of the above types)

See also upcoming "FieldSelectors" feature:
https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#field-selectors
https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#type

#### Preserve Unknown Fields

I like the current behaviour of v2 regarding this topic. It behaves as I would expect it.

> if we take the opinionated decision to add x-kubernetes-preserve-unknown everywhere by default ...

In my opinion using x-kubernetes-preserve-unknown everywhere could be "dangerous" because Kubernetes behaves differently. Removed fields in a CR won't get cleaned up (see [control-pruning](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#controlling-pruning))

shawkins commented:
> I would suggest to silently filter printer colums on other types. At the moment it produces an invalid CRD.

I'd be more in favor of producing a hard error or emitting a warning.

metacosm commented on crd-generator/api-v2/src/main/java/io/fabric8/crdv2/generator/CRDInfo.java:
```diff
@@ -0,0 +1,41 @@
+ */
+package io.fabric8.crdv2.generator;
+
+public class CRDInfo {
```
This is a no-go for QOSDK: this class is missing the `getDependentClassNames` support we need.

shawkins replied:
I can add this back, but just to double check - is there a quarkus utility for detecting the transitive graph of a given class?

shawkins replied:
Also - if this is needed, is it okay to just provide the dependent classes across everything passed into the crd generator? The ResolvingContext.uriToJacksonSchema values will contain all of the introspected classes, but it's not tracked by CRD.

metacosm replied:
> I can add this back, but just to double check - is there a quarkus utility for detecting the transitive graph of a given class?

I'm not sure if there's such a thing… I'll ask but, iirc, there wasn't one when I implemented this feature.

metacosm replied:
> > I can add this back, but just to double check - is there a quarkus utility for detecting the transitive graph of a given class?
> 
> I'm not sure if there's such a thing… I'll ask but, iirc, there wasn't one when I implemented this feature.

So I checked, and, no, there still isn't such a thing in Quarkus

metacosm replied:
> Also - if this is needed, is it okay to just provide the dependent classes across everything passed into the crd generator? The ResolvingContext.uriToJacksonSchema values will contain all of the introspected classes, but it's not tracked by CRD.

I'm not sure, I would need to look into things in deeper details. Obviously, this would probably mean more work on my side. Additionally, I'd rather not expose these internals to outside code, if possible. The `getDependentClassNames` method was implementation-agnostic so I think it would be useful to restore the functionality if possible.

shawkins replied:
> So I checked, and, no, there still isn't such a thing in Quarkus

Ok, and do you need it tracked per crd, or is across everything good enough?

shawkins replied:
@metacosm check now, it should map to what you were using before.

metacosm replied:
> @metacosm check now, it should map to what you were using before.

It doesn't work completely as expected. In particular, the previous version was overeager and returned too many things (like the `Group` annotation) but the new version doesn't return enough, in particular when there's a class hierarchy being involved (e.g. if a CR extends an intermediate class instead of extending CustomResource directly, the new version won't provide the intermediate class in `getDependentClassNames` even though changes to that class could potentially impact the CRD).

shawkins replied:
@metacosm check now. It's not filtering java level stuff, but it will get the interfaces and super classes as well.

metacosm replied:
Yep, this seems to work nicely now. Will now do a full CI run with this PR to check.

shawkins commented:
> > I would suggest to silently filter printer colums on other types. At the moment it produces an invalid CRD.
> 
> I'd be more in favor of producing a hard error or emitting a warning.

Added the warning and extracted common printer column handling.

manusa commented:
https://github.com/fabric8io/kubernetes-client/pull/5949#discussion_r1586456252

Is this ready to be merged then?

shawkins commented:
> Is this ready to be merged then?

I think so. We can certianly refine things from here. There's quite a few follow-on tasks that have already been logged and more that are mentioned in the PR description.

manusa reviewed (approved):
LGTM, thx!
